File: stream/RequestTransmitter.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package grakn.client.stream;

import grakn.client.common.GraknClientException;
import grakn.client.common.RequestBuilder;
import grakn.common.collection.ConcurrentSet;
import grakn.common.concurrent.NamedThreadFactory;
import grakn.protocol.TransactionProto;
import io.grpc.stub.StreamObserver;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.Semaphore;
import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.ReadWriteLock;
import java.util.concurrent.locks.StampedLock;

import static grakn.client.common.ErrorMessage.Client.CLIENT_CLOSED;

public class RequestTransmitter implements AutoCloseable {

    private static final Logger LOG = LoggerFactory.getLogger(RequestTransmitter.class);
    private static final int BATCH_WINDOW_SMALL_MILLIS = 1;
    private static final int BATCH_WINDOW_LARGE_MILLIS = 3;

    private final ArrayList<Executor> executors;
    private final AtomicInteger executorIndex;
    private final ReadWriteLock accessLock;
    private volatile boolean isOpen;

    public RequestTransmitter(int parallelisation, NamedThreadFactory threadFactory) {
        this.executors = new ArrayList<>(parallelisation);
        this.executorIndex = new AtomicInteger(0);
        this.accessLock = new StampedLock().asReadWriteLock();
        this.isOpen = true;
        for (int i = 0; i < parallelisation; i++) this.executors.add(new Executor(threadFactory));
    }

    private Executor nextExecutor() {
        return executors.get(executorIndex.getAndUpdate(i -> {
            i++;
            if (i % executors.size() == 0) i = 0;
            return i;
        }));
    }

    public Dispatcher dispatcher(StreamObserver<TransactionProto.Transaction.Client> requestObserver) {
        try {
            accessLock.readLock().lock();
            if (!isOpen) throw new GraknClientException(CLIENT_CLOSED);
            Executor executor = nextExecutor();
            Dispatcher dispatcher = new Dispatcher(executor, requestObserver);
            executor.dispatchers.add(dispatcher);
            return dispatcher;
        } finally {
            accessLock.readLock().unlock();
        }
    }

    @Override
    public void close() {
        try {
            accessLock.writeLock().lock();
            if (isOpen) {
                isOpen = false;
                executors.forEach(Executor::close);
            }
        } finally {
            accessLock.writeLock().unlock();
        }
    }

    private class Executor implements AutoCloseable {

        private final ConcurrentSet<Dispatcher> dispatchers;
        private final AtomicBoolean isRunning;
        private final Semaphore permissionToRun;

        private Executor(ThreadFactory threadFactory) {
            dispatchers = new ConcurrentSet<>();
            isRunning = new AtomicBoolean(false);
            permissionToRun = new Semaphore(0);
            threadFactory.newThread(this::run).start();
        }

        private void mayStartRunning() {
            if (isRunning.compareAndSet(false, true)) permissionToRun.release();
        }

        private void run() {
            while (isOpen) {
                try {
                    permissionToRun.acquire();
                    boolean first = true;
                    while (true) {
                        Thread.sleep(first ? BATCH_WINDOW_SMALL_MILLIS : BATCH_WINDOW_LARGE_MILLIS);
                        if (dispatchers.isEmpty()) break;
                        else dispatchers.forEach(Dispatcher::sendBatchedRequests);
                        first = false;
                    }
                } catch (InterruptedException e) {
                    LOG.error(e.getMessage(), e);
                } finally {
                    isRunning.set(false);
                }
                if (!dispatchers.isEmpty()) mayStartRunning();
            }
        }

        @Override
        public void close() {
            dispatchers.forEach(Dispatcher::close);
            mayStartRunning();
        }
    }

    public class Dispatcher implements AutoCloseable {

        private final Executor executor;
        private final StreamObserver<TransactionProto.Transaction.Client> requestObserver;
        private final ConcurrentLinkedQueue<TransactionProto.Transaction.Req> requestQueue;

        private Dispatcher(Executor executor, StreamObserver<TransactionProto.Transaction.Client> requestObserver) {
            this.executor = executor;
            this.requestObserver = requestObserver;
            requestQueue = new ConcurrentLinkedQueue<>();
        }

        private synchronized void sendBatchedRequests() {
            if (requestQueue.isEmpty() || !isOpen) return;
            TransactionProto.Transaction.Req request;
            ArrayList<TransactionProto.Transaction.Req> requests = new ArrayList<>(requestQueue.size() * 2);
            while ((request = requestQueue.poll()) != null) requests.add(request);
            requestObserver.onNext(RequestBuilder.Transaction.clientMsg(requests));
        }

        public void dispatch(TransactionProto.Transaction.Req requestProto) {
            try {
                accessLock.readLock().lock();
                if (!isOpen) throw new GraknClientException(CLIENT_CLOSED);
                requestQueue.add(requestProto);
                executor.mayStartRunning();
            } finally {
                accessLock.readLock().unlock();
            }
        }

        public void dispatchNow(TransactionProto.Transaction.Req requestProto) {
            try {
                accessLock.readLock().lock();
                if (!isOpen) throw new GraknClientException(CLIENT_CLOSED);
                requestQueue.add(requestProto);
                sendBatchedRequests();
            } finally {
                accessLock.readLock().unlock();
            }
        }

        @Override
        public void close() {
            requestObserver.onCompleted();
            executor.dispatchers.remove(this);
        }
    }
}

File: stream/RequestTransmitter.ts
/*
 * Copyright (C) 2021 Vaticle
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import {ClientDuplexStream} from "@grpc/grpc-js";
import {Transaction as TransactionProto} from "typedb-protocol/common/transaction_pb";
import {RequestBuilder} from "../common/rpc/RequestBuilder";

export class BatchDispatcher {

    private static BATCH_WINDOW_SMALL_MILLIS = 1;
    private static BATCH_WINDOW_LARGE_MILLIS = 3;

    private readonly _transmitter: RequestTransmitter;
    private readonly _transactionStream: ClientDuplexStream<TransactionProto.Client, TransactionProto.Server>;
    private _bufferedRequests: TransactionProto.Req[];
    private _isRunning: boolean;

    constructor(transmitter: RequestTransmitter, transactionStream: ClientDuplexStream<TransactionProto.Client, TransactionProto.Server>) {
        this._transmitter = transmitter;
        this._transactionStream = transactionStream;
        this._bufferedRequests = new Array<TransactionProto.Req>();
        this._isRunning = false;
    }

    dispatch(req: TransactionProto.Req): void {
        this._bufferedRequests.push(req);
        this.sendScheduledBatch();
    }

    dispatchNow(req: TransactionProto.Req): void {
        this._bufferedRequests.push(req);
        this.sendNow();
    }

    close(): void {
        this._transmitter._dispatchers.delete(this);
        this._transactionStream.end();
    }

    private sendNow(): void {
        const clientRequest = RequestBuilder.Transaction.clientReq(this._bufferedRequests);
        this._transactionStream.write(clientRequest);
        this._bufferedRequests = [];
    }

    private sendScheduledBatch(): void {
        if (this._isRunning) return;
        this._isRunning = true;
        this.setSchedule(true);
    }

    private setSchedule(first: boolean) {
        const wait = first ? BatchDispatcher.BATCH_WINDOW_SMALL_MILLIS : BatchDispatcher.BATCH_WINDOW_LARGE_MILLIS;
        setTimeout(() => {
            if (this._bufferedRequests.length > 0) {
                this.sendNow();
                this.setSchedule(false);
            } else {
                this._isRunning = false;
            }
        }, wait);
    }
}

export class RequestTransmitter {

    readonly _dispatchers: Set<BatchDispatcher>;
    private _isOpen: boolean;

    constructor() {
        this._dispatchers = new Set<BatchDispatcher>();
        this._isOpen = true;
    }

    close(): void {
        if (this._isOpen) {
            this._isOpen = false;
            this._dispatchers.forEach(dispatcher => dispatcher.close());
        }
    }

    dispatcher(transactionStream: ClientDuplexStream<TransactionProto.Client, TransactionProto.Server>): BatchDispatcher {
        const dispatcher = new BatchDispatcher(this, transactionStream);
        this._dispatchers.add(dispatcher);
        return dispatcher;
    }
}
